contracts/interchain-token-service/src/contract/execute/interceptors.rs

```diff
@@ -0,0 +1,648 @@
+use axelar_wasm_std::{nonempty, FnExt};
+use cosmwasm_std::{Storage, Uint256};
+use error_stack::{bail, ensure, report, Result, ResultExt};
+use router_api::ChainNameRaw;
+
+use super::Error;
+use crate::state::{self, TokenDeploymentType};
+use crate::{DeployInterchainToken, InterchainTransfer, TokenConfig, TokenId, TokenInstance};
+
+pub fn subtract_supply_amount(
+    storage: &mut dyn Storage,
+    chain: &ChainNameRaw,
+    transfer: &InterchainTransfer,
+) -> Result<(), Error> {
+    let mut token = try_load_token_instance(storage, chain.clone(), transfer.token_id)?;
+
+    token.supply = token
+        .supply
+        .checked_sub(transfer.amount)
+        .change_context_lazy(|| Error::TokenSupplyInvariantViolated {
+            token_id: transfer.token_id,
+            chain: chain.clone(),
+        })?;
+
+    state::save_token_instance(storage, chain.clone(), transfer.token_id, &token)
+        .change_context(Error::State)
+}
+
+pub fn add_supply_amount(
+    storage: &mut dyn Storage,
+    chain: &ChainNameRaw,
+    transfer: &InterchainTransfer,
+) -> Result<(), Error> {
+    let mut token = try_load_token_instance(storage, chain.clone(), transfer.token_id)?;
+
+    token.supply = token
+        .supply
+        .checked_add(transfer.amount)
+        .change_context_lazy(|| Error::TokenSupplyInvariantViolated {
+            token_id: transfer.token_id,
+            chain: chain.clone(),
+        })?;
+
+    state::save_token_instance(storage, chain.clone(), transfer.token_id, &token)
+        .change_context(Error::State)
+}
+
+pub fn apply_scaling_factor_to_amount(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    destination_chain: &ChainNameRaw,
+    transfer: &mut InterchainTransfer,
```

Comment:
just return this, instead of mutating. The function doesn't return anything as is